frontend/src/main/scala/bloop/engine/tasks/ScalaNative.scala

```diff
@@ -0,0 +1,130 @@
+package bloop.engine.tasks
+
+import scala.util.{Failure, Success, Try}
+
+import java.net.URLClassLoader
+import java.nio.file.{Files, Path}
+import java.util.concurrent.ConcurrentHashMap
+
+import bloop.{DependencyResolution, Project}
+import bloop.cli.{ExitStatus, OptimizerConfig}
+import bloop.engine.State
+import bloop.exec.Forker
+import bloop.io.AbsolutePath
+import bloop.logging.Logger
+
+import monix.eval.Task
+
+object ScalaNative {
+
+  private[this] var _ivyResolved: ScalaNative = _
+  private[this] val instancesCache: ConcurrentHashMap[Array[AbsolutePath], ScalaNative] =
+    new ConcurrentHashMap
+
+  def forProject(project: Project, logger: Logger): ScalaNative = {
+    project.nativeConfig match {
+      case None =>
+        ivyResolved(logger)
+      case Some(config) =>
+        val classpath = config.toolchainClasspath.map(AbsolutePath.apply)
+        direct(classpath)
+    }
+  }
+
+  def ivyResolved(logger: Logger): ScalaNative = synchronized {
+    if (_ivyResolved == null) {
```

Comment:
What is the purpose of `_ivyResolved` in the grand scheme of things? Don't we already cache with `instancesCache` and synchronize on the object here?

Reply:
This is only to completely bypass coursier (though I remember measuring it when we discussed about resolving the test agents, and it took only a few ms). I'm happy to get rid of it.

Reply:
Just replied with another comment re caching.